keps/sig-network/20201010-kube-proxy-v2/README.md

```diff
@@ -0,0 +1,753 @@
+
+This KEP is born from the conviction that more decoupling of the Service object and the actual implementations is required,
+by introducing an intermediate, node-level abstaction provider. This abstraction is expected to be the result of applying
+Kubernetes' `Service` semantics and business logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business logic changes
+- provide a client library with minimal dependencies
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- (optional) help proxy implementations using the same subsystem (ie iptables) to cooperate more easily
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- provide equivalent implementations of third-party ones
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible via classic TCP (`127.0.0.1:12345`)
+and/or a socket (`unix:///path/to/proxy.sock`).
+
+- it will connect to the API server and watch resources, like the current proxy;
+- then, it will process them, applying Kubernetes specific business logic like topology computation relative to the local host;
+- finally, provide the result of this computation to client via a gRPC watchable API.
+
+The idea is to sent the full state to the client, so implementations don't have to do diff-processing and maintain any internal state.
+This should provide simple implementations, reliable results and still be quite optimal, since many kernel network-level objects are
+updated via atomic replace APIs. It's also a protection from slow readers, since no stream has to be buffered.
+
+Since the node-local state computed by the new proxy will be simpler and node-specific, it will only change when the result for the
+current node is actually changed. Since there's less data in the local state, change frequency is reduced. Testing on actual clusters
```

Comment:
With some network plugins, there is no per-node kube-proxy equivalent, because Services are just a specific case of the general cluster-wide network programming that is handled by some other system. Eg, in the case of ovn-kubernetes, the single ovn-kubernetes master controller watches Service and Endpoint changes and then updates the OVN database to reflect those changes. The OVN DB changes are then propagated to the individual nodes as needed, where `ovn-controller` processes make the necessary routing/filtering/etc changes for that node.

So, ovn-kubernetes could benefit from a helper that processed some of the Service-related information for it, but it would only want a single helper that gave it a top-level view of Services, not a per-node helper that gave a node-level view...

Reply:
yes, it is expected that not every use-case will benefit from this, but will still benefit from the simpler and more stable API.